File: remote-ws/lib/remote_ws/telemetry.ex
defmodule RemoteWs.Telemetry do
  @moduledoc """
  OpenTelemetry for the relay: traces, metrics and logs over OTLP.

  Mirrors what the Rust services get from `rocksky-telemetry`, the Go ones from
  `otel/` and the Node ones from `apps/api/src/otel.ts` — the same collector
  (`OTEL_EXPORTER_OTLP_ENDPOINT`, `OTEL_EXPORTER_OTLP_HEADERS`) and the same
  instrument names, so all of Rocksky lands in one set of charts.

  ## Why three application environments

  Traces are a stable signal in the Erlang SDK and configure under
  `:opentelemetry` / `:opentelemetry_exporter`. Metrics and logs are not: the
  meter provider, `otel_metric_reader` and the `otel_log_handler` logger handler
  all live in `:opentelemetry_experimental`, and their OTLP exporters read the
  endpoint out of *that* application's environment. That is why
  config/runtime.exs sets the same endpoint under two keys rather than one.

  ## What is instrumented

    * HTTP — Bandit and Phoenix, so `/health` and the `/ws` upgrade each get a
      span. The upgrade's span ends at the 101; it does not stay open for the
      life of the socket.
    * Postgres and Redis — every `Ecto` query and `Redix` command, as children
      of whatever relay span was open.
    * The relay itself — one span per inbound frame (`RemoteWs.Ws.Handler`), one
      per disconnect, one per debounced `song.stopped`, and one per NATS
      publish.

  Logs keep going through `Logger`; the handler installed here ships a copy to
  the collector stamped with the trace and span id of whatever span was open
  when the line was written, so a log line links back to the frame that produced
  it.

  ## What ties the three signals together

    * One resource on all three. `otel_resource_detector:get_resource/0` is what
      the tracer, the metric reader and `otel_log_handler` each read, so the
      `service.name`, `service.version`, `service.namespace` and
      `service.instance.id` set in config/config.exs and config/runtime.exs are
      on every span, metric point and log record alike.
    * Trace and span ids on the logs, as above.
    * The trace crosses the process boundary: the SDK's default
      `otel_propagator_trace_context` reads the W3C `traceparent` off the HTTP
      upgrade and writes it onto outgoing requests, which is the same format
      the Rust, Go, Node and Deno services use.
    * The same instrument names and attribute keys as the rest of the fleet —
      see `RemoteWs.Telemetry.Metrics` — so one chart covers every service and
      its filter narrows the traces behind a spike.
    * Exemplars: the trace id of a sampled span attached to the histogram
      bucket its measurement landed in, so a spike on a chart clicks straight
      through to one of the frames behind it. Switched on in
      config/runtime.exs, and it works because `RemoteWs.Telemetry.Metrics`
      records through `:otel_ctx.get_current()` — a measurement taken with an
      empty context is never an exemplar.

  Exemplars are the one link that is not available fleet-wide: the Go services
  export them and this one does, the Rust, Node and Deno SDKs cannot produce
  them at all. On those, the route from a metric to a trace is the shared
  attribute names above.
  """

  require Logger
  require OpenTelemetry.Tracer

  alias RemoteWs.Telemetry.Metrics

  @doc """
  Attaches every instrumentation handler and installs the log handler. Called
  once from `RemoteWs.Application.start/2`, before the endpoint starts listening.
  """
  def setup do
    OpentelemetryBandit.setup()
    OpentelemetryPhoenix.setup(adapter: :bandit, liveview: false)
    OpentelemetryEcto.setup([:remote_ws, :repo])
    OpentelemetryRedix.setup()

    Metrics.setup()
    attach_http_metrics()

    if Application.get_env(:remote_ws, :otlp_logs, true), do: install_log_handler()

    :ok
  end

  @doc """
  Runs `fun` as one unit of relay work: a root span named `name`, plus a sample
  on `rocksky.work.items` / `rocksky.work.duration` tagged with the same name.

  A root span rather than a child, because a frame arrives on a socket that was
  established long ago: there is no incoming `traceparent` to continue. The
  context is cleared first — the connection process is the same one that served
  the HTTP upgrade, and whatever it left behind would otherwise parent every
  frame for the life of the socket onto one enormous trace.
  """
  def work(name, attributes, fun) do
    :otel_ctx.clear()
    started = System.monotonic_time()

    OpenTelemetry.Tracer.with_span name, %{attributes: attributes} do
      try do
        result = fun.()
        Metrics.record_work(name, "ok", elapsed_seconds(started))
        result
      catch
        kind, reason ->
          fail(kind, reason)
          Metrics.record_work(name, "error", elapsed_seconds(started))
          :erlang.raise(kind, reason, __STACKTRACE__)
      end
    end
  end

  @doc """
  Runs `fun` inside a child span of whatever is currently open — for the
  outward effects (a NATS publish) that hang off a unit of work rather than
  being one.
  """
  def span(name, attributes, fun) do
    OpenTelemetry.Tracer.with_span name, %{attributes: attributes} do
      try do
        fun.()
      catch
        kind, reason ->
          fail(kind, reason)
          :erlang.raise(kind, reason, __STACKTRACE__)
      end
    end
  end

  defp fail(kind, reason) do
    OpenTelemetry.Tracer.set_status(OpenTelemetry.status(:error, "#{kind}: #{inspect(reason)}"))
  end

  defp elapsed_seconds(started) do
    System.convert_time_unit(System.monotonic_time() - started, :native, :nanosecond) /
      1_000_000_000
  end

  # ── HTTP metrics ────────────────────────────────────────────────────────────
  #
  # Off Bandit's events rather than Phoenix's: `Plug.Telemetry`'s `:stop` fires
  # from a `before_send` hook, and the `/ws` upgrade never sends a response
  # through one — so measuring there would report on `/health` and nothing else.

  defp attach_http_metrics do
    :telemetry.attach_many(
      {__MODULE__, :http_metrics},
      [[:bandit, :request, :stop], [:bandit, :request, :exception]],
      &__MODULE__.handle_http_event/4,
      :no_config
    )
  end

  @doc false
  def handle_http_event([:bandit, :request, :stop], measurements, %{conn: conn}, _config) do
    record_request(conn, conn.status || 200, measurements)
  end

  def handle_http_event([:bandit, :request, :exception], measurements, meta, _config) do
    case meta do
      %{conn: conn} -> record_request(conn, 500, measurements)
      _ -> :ok
    end
  end

  def handle_http_event(_event, _measurements, _meta, _config), do: :ok

  defp record_request(conn, status, measurements) do
    seconds =
      System.convert_time_unit(measurements[:duration] || 0, :native, :nanosecond) / 1_000_000_000

    Metrics.record_request(route(conn), conn.method, status, seconds)
  end

  # By route, not by path: this service only serves `/ws` and `/health` today,
  # but labelling by the raw path is how a metric quietly grows a time series
  # per URL the day something else is mounted here.
  defp route(conn) do
    case Phoenix.Router.route_info(RemoteWsWeb.Router, conn.method, conn.request_path, conn.host) do
      %{route: route} -> route
      _ -> "unmatched"
    end
  end

  # ── logs ────────────────────────────────────────────────────────────────────

  defp install_log_handler do
    config = %{
      level: :info,
      filter_default: :log,
      filters: [otel_internal: {&__MODULE__.drop_otel_internal/2, []}],
      exporter: {:otel_exporter_logs_otlp, %{}}
    }

    case :logger.add_handler(:otel_log_handler, :otel_log_handler, config) do
      :ok -> :ok
      {:error, {:already_exist, _}} -> :ok
      {:error, reason} -> Logger.error("otel: log exporter disabled: #{inspect(reason)}")
    end
  end

  @doc false
  # The SDK logs its own export failures. Shipping those to the collector is how
  # a collector that is refusing everything turns into a loop that generates
  # more of exactly what it cannot deliver, so they stay on the console only.
  def drop_otel_internal(%{meta: %{mfa: {module, _function, _arity}}} = event, _extra) do
    if otel_module?(module), do: :stop, else: event
  end

  def drop_otel_internal(event, _extra), do: event

  defp otel_module?(module) do
    name = Atom.to_string(module)
    String.starts_with?(name, "otel_") or String.starts_with?(name, "opentelemetry")
  end
end

File: apps/feeds/src/telemetry.ts
import {
  metrics,
  type Span,
  SpanKind,
  SpanStatusCode,
  trace,
} from "@opentelemetry/api";
import type { MiddlewareHandler } from "@hono/hono";
import type { Pool } from "pg";

// Deno ships OpenTelemetry in the runtime: with OTEL_DENO=true it exports
// traces for Deno.serve/fetch, the runtime metrics, and every console.* call as
// a log record, all over OTLP. There is no SDK to start here — this module only
// adds the spans and metrics Deno cannot know about.
//
// What ties the three signals together:
//
//   * the resource, which Deno builds from the environment before any of this
//     module runs and so cannot be set here — OTEL_SERVICE_NAME and
//     OTEL_RESOURCE_ATTRIBUTES are in the deno.json tasks for that reason.
//     Without them every signal exports as "unknown_service" and none of them
//     can be tied to the others or to the rest of the fleet.
//   * trace and span ids on every log line, which the runtime attaches to each
//     console.* record from the span that is open at the time — so the logtape
//     sink in context.ts is also the log exporter, and its output stays plain
//     when telemetry is on.
//   * W3C traceparent in and out: Deno.serve continues an incoming trace and
//     fetch injects into an outgoing one, so a feed request that starts in the
//     appview shows as one trace across both.
//
// Exemplars — a trace id on one histogram sample, so a chart clicks through to
// the slow request behind a spike — the runtime does not implement, and there
// is no SDK here to configure one on. Getting from a metric to a trace means
// filtering the traces by the attributes the metric carries.

export const SERVICE_NAME = "rocksky-feeds";

export const tracer = trace.getTracer(SERVICE_NAME);
const meter = metrics.getMeter(SERVICE_NAME);

const httpRequestCounter = meter.createCounter("http_requests_total", {
  description: "Count of incoming requests, by route and status code",
});

const httpRequestDuration = meter.createHistogram(
  "http_request_duration_seconds",
  {
    description: "Request duration in seconds, by route",
    unit: "s",
  },
);

export const feedRequestCounter = meter.createCounter("feed_requests_total", {
  description: "Count of feed skeleton requests, by algorithm and outcome",
});

export const feedRequestDuration = meter.createHistogram(
  "feed_request_duration_seconds",
  {
    description: "Feed skeleton duration in seconds, by algorithm",
    unit: "s",
  },
);

export const feedItemsReturned = meter.createHistogram("feed_items_returned", {
  description: "Number of scrobbles returned per feed skeleton request",
});

const dbQueryDuration = meter.createHistogram("db_query_duration_seconds", {
  description: "Postgres query duration in seconds, by operation",
  unit: "s",
});

const seconds = (startedAt: number) => (performance.now() - startedAt) / 1000;

export function recordError(span: Span, error: unknown): void {
  span.recordException(error instanceof Error ? error : String(error));
  span.setStatus({
    code: SpanStatusCode.ERROR,
    message: error instanceof Error ? error.message : String(error),
  });
}

// The NSID is the only useful route label here: the XRPC router matches every
// method on one pattern, so c.req.routePath would collapse them all together.
const routeOf = (path: string): string =>
  path.startsWith("/xrpc/") ? path.slice("/xrpc/".length) : path;

export function telemetry(): MiddlewareHandler {
  return async (c, next) => {
    const route = routeOf(c.req.path);
    const startedAt = performance.now();
    trace.getActiveSpan()?.setAttributes({
      "http.route": route,
      ...(route === c.req.path ? {} : { "xrpc.method": route }),
    });

    let status = 500;
    try {
      await next();
      status = c.res.status;
    } finally {
      const labels = { route, method: c.req.method, status };
      httpRequestCounter.add(1, labels);
      httpRequestDuration.record(seconds(startedAt), labels);
    }
  };
}

type QueryArgs = readonly unknown[];
type QueryFn = (...args: QueryArgs) => unknown;

const statementOf = (config: unknown): string => {
  if (typeof config === "string") return config;
  if (config && typeof config === "object" && "text" in config) {
    return String((config as { text: unknown }).text);
  }
  return "";
};

// Drizzle runs every query through pool.query, so wrapping it once covers the
// whole data layer. Statement text is parameterized by drizzle; the bound
// values stay out of telemetry.
export function instrumentPool(pool: Pool): Pool {
  const query = pool.query.bind(pool) as QueryFn;

  const instrumented: QueryFn = (...args: QueryArgs) => {
    // The callback form never returns a promise; leave it alone.
    if (typeof args.at(-1) === "function") return query(...args);

    const statement = statementOf(args[0]);
    const operation = /^\s*(\w+)/.exec(statement)?.[1]?.toUpperCase() ??
      "UNKNOWN";
    const startedAt = performance.now();

    return tracer.startActiveSpan(
      "pg.query",
      {
        kind: SpanKind.CLIENT,
        attributes: {
          "db.system": "postgresql",
          "db.operation": operation,
          "db.statement": statement,
        },
      },
      (span) => {
        const finish = (status: "ok" | "error") => {
          span.end();
          dbQueryDuration.record(seconds(startedAt), { operation, status });
        };
        try {
          return Promise.resolve(query(...args)).then(
            (result) => {
              finish("ok");
              return result;
            },
            (error: unknown) => {
              recordError(span, error);
              finish("error");
              throw error;
            },
          );
        } catch (error) {
          recordError(span, error);
          finish("error");
          throw error;
        }
      },
    );
  };

  pool.query = instrumented as typeof pool.query;
  return pool;
}
